test/e2e/info_api_test.go

```diff
@@ -0,0 +1,178 @@
+	testutil.Ok(t, err)
+	prom3, sidecar3, err := e2ethanos.NewPrometheusWithSidecar(s.SharedDir(), "e2e_test_info", "alone3", defaultPromConfig("prom-alone3", 0, "", ""), e2ethanos.DefaultPrometheusImage())
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2, prom3, sidecar3))
+
+	m := e2edb.NewMinio(8080, "thanos")
+	testutil.Ok(t, s.StartAndWaitReady(m))
+	str, err := e2ethanos.NewStoreGW(s.SharedDir(), "1", client.BucketConfig{
+		Type: client.S3,
+		Config: s3.Config{
+			Bucket:    "thanos",
+			AccessKey: e2edb.MinioAccessKey,
+			SecretKey: e2edb.MinioSecretKey,
+			Endpoint:  m.NetworkHTTPEndpoint(),
+			Insecure:  true,
+		},
+	})
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(str))
+
+	// Register `sidecar1` in all flags (i.e. '--store', '--rule', '--target', '--metadata', '--exemplar', '--endpoint') to verify
+	// '--endpoint' flag works properly works together with other flags ('--target', '--metadata' etc.).
+	// Register 2 sidecars and 1 storeGW using '--endpoint'.
+	// Register `sidecar3` twice to verify it is deduplicated.
+	q, err := e2ethanos.NewQuerierBuilder(s.SharedDir(), "1", []string{sidecar1.GRPCNetworkEndpoint()}).
+		WithTargetAddresses([]string{sidecar1.GRPCNetworkEndpoint()}).
+		WithMetadataAddresses([]string{sidecar1.GRPCNetworkEndpoint()}).
+		WithExemplarAddresses([]string{sidecar1.GRPCNetworkEndpoint()}).
+		WithRuleAddresses([]string{sidecar1.GRPCNetworkEndpoint()}).
+		WithEndpoints([]string{
+			sidecar1.GRPCNetworkEndpoint(),
+			sidecar2.GRPCNetworkEndpoint(),
+			sidecar3.GRPCNetworkEndpoint(),
+			sidecar3.GRPCNetworkEndpoint(),
+			str.GRPCNetworkEndpoint(),
+		}).
+		Build()
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(q))
+
+	expected := map[string][]query.StoreStatus{
+		"sidecar": {
+			{
+				Name: "e2e_test_info-sidecar-alone1:9091",
+				LabelSets: []labels.Labels{{
+					{
+						Name:  "prometheus",
+						Value: "prom-alone1",
+					},
+					{
+						Name:  "replica",
+						Value: "0",
+					},
+				}},
+			},
+			{
+				Name: "e2e_test_info-sidecar-alone2:9091",
+				LabelSets: []labels.Labels{{
+					{
+						Name:  "prometheus",
+						Value: "prom-alone2",
+					},
+					{
+						Name:  "replica",
+						Value: "0",
+					},
+				}},
+			},
+			{
+				Name: "e2e_test_info-sidecar-alone3:9091",
+				LabelSets: []labels.Labels{{
+					{
+						Name:  "prometheus",
+						Value: "prom-alone3",
+					},
+					{
+						Name:  "replica",
+						Value: "0",
+					},
+				}},
+			},
+		},
+		"store": {
+			{
+				Name:      "e2e_test_info-store-gw-1:9091",
+				LabelSets: []labels.Labels{},
+			},
+		},
+	}
+
+	url := "http://" + path.Join(q.HTTPEndpoint(), "/api/v1/stores")
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+	defer cancel()
+
+	err = runutil.Retry(time.Second, ctx.Done(), func() error {
+
+		resp, err := http.Get(url)
```

Comment:
Nit: maybe we could call this with `ctx`? There could be problems if this hangs indefinitely for some reason